sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureFileSystemProvider.java

```diff
@@ -3,9 +3,299 @@
+ * containers may be specified as file stores upon creation of the file system. When a file system is created,
+ * it will try to retrieve the properties of each container to ensure connection to the account. If any of the
+ * containers does not exist, it will be created. Failure to access or create containers as necessary will result in
+ * an exception and failure to create the file system. Any data existing in the containers will be preserved and
+ * accessible via the file system, though customers should be aware that it must be in a format understandable by
+ * the types in this package or behavior will be undefined.
+ * <p>
+ * {@link #newFileSystem(URI, Map)} will check for the following keys in the configuration map and expect the named
+ * types. Any entries not listed here will be ignored. Note that {@link AzureFileSystem} has public constants defined
+ * for each of the keys for convenience.
+ * <ul>
+ *     <li>{@code AzureStorageAccountKey:}{@link String}</li>
+ *     <li>{@code AzureStorageSasToken:}{@link String}</li>
+ *     <li>{@code AzureStorageHttpLogDetailLevel:}{@link com.azure.core.http.policy.HttpLogDetailLevel}</li>
+ *     <li>{@code AzureStorageMaxTries:}{@link Integer}</li>
+ *     <li>{@code AzureStorageTryTimeout:}{@link Integer}</li>
+ *     <li>{@code AzureStorageRetryDelayInMs:}{@link Long}</li>
+ *     <li>{@code AzureStorageMaxRetryDelayInMs:}{@link Long}</li>
+ *     <li>{@code AzureStorageRetryPolicyType:}{@link com.azure.storage.common.policy.RetryPolicyType}</li>
+ *     <li>{@code AzureStorageSecondaryHost:}{@link String}</li>
+ *     <li>{@code AzureStorageSecondaryHost:}{@link Integer}</li>
+ *     <li>{@code AzureStorageBlockSize:}{@link Integer}</li>
+ *     <li>{@code AzureStorageDownloadResumeRetries:}{@link Integer}</li>
+ *     <li>{@code AzureStorageUseHttps:}{@link Boolean}</li>
+ *     <li>{@code AzureStorageFileStores:}{@link Iterable<String>}</li>
+ * </ul>
+ * <p>
+ * Either an account key or a sas token must be specified. If both are provided, the account key will be preferred. If
+ * a sas token is specified, the customer must take care that it has appropriate permissions to perform the actions
+ * demanded of the file system in a given workflow, including the initial connection check specified above. Furthermore,
+ * it must have an expiry time that lasts at least until the file system is closed as there is no token refresh offered
+ * at this time. The same token will be applied to all containers.
+ * <p>
+ * An iterable of file stores must also be provided; each entry should simply be the name of a container. All other
+ * values listed are used to configure the underlying {@link com.azure.storage.blob.BlobServiceClient}. Please refer to
+ * that type for more information on these values.
+ *
+ * @see FileSystemProvider
  */
-public class AzureFileSystemProvider  {
+public final class AzureFileSystemProvider extends FileSystemProvider {
+    // TODO: Add logger
```

Comment:
@alzimmermsft Do you think we should use the same system for logging exceptions as in the SDKs. In particular, I think some of the logging methods return RuntimeExceptions, but the spec dictates that specific types of exceptions need to be thrown here. But we probably still should capture a lot of these errors in logs.

Reply:
Should log these exceptions but don't need to throw them using `ClientLogger.logExceptionAs{Warning/Error}`, instead could call `ClientLogger.warning(Message, Exception)` and `ClientLogger.error(Message, Exception)` and throw the non-runtime exception afterwards. This technique is used in a few places where non-runtime exceptions are being thrown.

Reply:
Will do. Thanks